pkg/deploy/controller/deployment/controller.go

```diff
@@ -136,8 +137,55 @@ func (c *DeploymentController) Handle(deployment *kapi.ReplicationController) er
 		}
 		c.recorder.Eventf(deployment, "cancelled", "Cancelled deployment")
 	case deployapi.DeploymentStatusFailed:
-		// Nothing to do in this terminal state.
-		glog.V(4).Infof("Ignoring deployment %s (status %s)", deployutil.LabelForDeployment(deployment), currentStatus)
+		// Scale down the current failed deployment
+		configName := deployutil.DeploymentConfigNameFor(deployment)
+		existingDeployments, err := c.deploymentClient.listDeploymentsForConfig(deployment.Namespace, configName)
+		if err != nil {
+			return fmt.Errorf("couldn't list Deployments for DeploymentConfig %s: %v", configName, err)
+		}
+
+		if len(existingDeployments.Items) > 0 {
+			var foundCurrentDeployment, success bool
+			var desiredReplicas int
+			sort.Sort(deployutil.DeploymentsByLatestVersionDesc(existingDeployments.Items))
+			for _, existing := range existingDeployments.Items {
+				// loop back through the deployments till you have crossed the current deployment
+				if existing.Name == deployment.Name {
+					foundCurrentDeployment = true
+					desiredReplicas, success = deployutil.DeploymentDesiredReplicas(&existing)
+					if !success {
+						return fmt.Errorf("Could not determine desired replicas from %s to reset replicas for last completed deployment", deployutil.LabelForDeployment(&existing))
+					}
+					continue
+				}
+				if !foundCurrentDeployment {
+					continue
+				}
+				// the first completed deployment is the one that needs to be scaled up
+				if deployutil.DeploymentStatusFor(&existing) == deployapi.DeploymentStatusComplete {
+					if existing.Spec.Replicas == desiredReplicas {
+						break
+					}
+					existing.Spec.Replicas = desiredReplicas
+					if _, err := c.deploymentClient.updateDeployment(existing.Namespace, &existing); err != nil {
+						c.recorder.Eventf(&existing, "failedUpdate", "Error trying to reset replicas to %d for deployment %s", desiredReplicas, deployutil.LabelForDeployment(&existing))
+						return fmt.Errorf("couldn't reset replicas to %d for deployment %s: %v", desiredReplicas, deployutil.LabelForDeployment(&existing), err)
+					}
+					glog.V(4).Infof("Updated replicas to %d for deployment %s", desiredReplicas, deployutil.LabelForDeployment(&existing))
+					break
+				}
+			}
+		}
+		// set the replicas for the failed deployment to 0
+		if deployment.Spec.Replicas == 0 {
+			break
+		}
+		deployment.Spec.Replicas = 0
+		if _, err := c.deploymentClient.updateDeployment(deployment.Namespace, deployment); err != nil {
```

Comment:
Since I ported the rest of the code to use the `kubectl.Scaler` interface, maybe we should use it here too.